xxxx-ledger-plugin-interface-2/xxxx-ledger-plugin-interface-2.md

```diff
@@ -0,0 +1,502 @@
+|:--|:--|:--|
+| `String` | [currencyCode](#ledgerinfocurrencycode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
+| `Number` | [currencyScale](#ledgerinfocurrencyscale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
+
+### Fields
+
+#### LedgerInfo#currencyCode
+<code>**currencyCode**:String</code>
+
+The [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) currency code (if any) used by the ledger. A custom all-caps three-letter code, not used by ISO 4217, otherwise.
+Ledger administrators who choose a custom currency code MAY request a custom currency symbol for their chosen currency code be listed by software modules that map currency codes to currency symbols,
+for instance on node package manager (npm) in the case of JavaScript.
+To translate an integer amount or balance from the ledger, the currencyCode by itself is not enough. It has to be used in combination with the currencyScale (below) to determine how many
+of the ledger's base units correspond to one currency unit.
+
+#### LedgerInfo#currencyScale
+<code>**currencyScale**:String</code>
+
+The order of magnitude to express one full currency unit in ledger's base units. For instance, if the integer values represented on the ledger are to be interpreted as
+dollar-cents (for the purpose of settling a user's account balance, for instance), then the ledger's
+currencyCode is `USD` and its currencyScale is `2`.
+
+## Class: PluginOptions
+<code>class PluginOptions</code>
+
+Plugin options are passed in to the [`LedgerPlugin`](#class-ledgerplugin)
+constructor when a plugin is being instantiated. The fields are ledger
+specific. Any fields which cannot be represented as strings are preceded with
+an underscore, and listed in the table below.
+
+###### Special Fields
+| Type | Name | Description |
+|:--|:--|:--|
+| `Object` | [_store](#pluginoptions-_store) | Persistence layer callbacks |
+
+### Fields
+
+#### PluginOptions#_store
+<code>**_store**:Object</code>
+
+Provides callback hooks to the host's persistence layer.
+
+Persistence MAY be required for internal use by some ledger plugins. For this purpose hosts MAY be configured with a persistence layer.
+
+Method names are based on the popular LevelUP/LevelDOWN packages.
+
+###### Example
+```js
+{
+  // Store a value under a key
+  put: (key, value) => {
+    // Returns Promise.<null>
+  },
+  // Fetch a value by key
+  get: (key) => {
+    // Returns Promise.<Object>
+  },
+  // Delete a value by key
+  del: (key) => {
+    // Returns Promise.<null>
+  }
+}
+```
+
+## Class: ConnectOptions
+<code>class ConnectOptions</code>
+
+###### Fields
+| Type | Name | Description |
+|:--|:--|:--|
+| `Number` | [timeout](#connectoptions-timeout) | milliseconds |
+
+### Fields
+
+#### ConnectOptions#timeout
+<code>**timeout**:Number</code>
+
+The number of milliseconds that the plugin should spend trying to connect before giving up.
```

Comment:
Couldn't this just be handled by the caller (with a `Promise.race`)?

Reply:
Hmm, how would you tell the plugin to stop trying to connect? Call `disconnect`?

Reply:
Ironically, I just ran into an issue in `ilp` that's very similar to this. `ilp` uses `Promise.race` to timeout connecting, but `Promise.race` doesn't do anything to stop the thing that "lost" the race.

In my case what this did is that the connection timeout was keeping the tests from exiting. (Mocha 4 no longer force-exits unlike Mocha 3, so you have to clean everything up properly.)

I did a quick fix by setting a shorter timeout. But imo, the proper fix is to remove `Promise.race` and replace it with `setTimeout` and call `clearTimeout` once the connection completes successfully. If you really wanted to use promises, you could use bluebird and it's idea of cancellation, but I think that's actually less clean.

General lesson: When using `Promise.race`, be aware that the thing that loses the race may keep doing stuff that's no longer necessary and/or that you don't want it to be doing.